### PR TITLE
Add minimal bucket DB read snapshot guard interface

### PR DESCRIPTION
**Pull request by vekterli**
@geirst please review Only exposes enough functionality to be used for Get operations for now.

* Enable free-lists for underlying replica `ArrayStore`.
* Legacy `MapBucketDatabase` read guard is _not_ thread safe, as it will never be used for non-blocking reads.
* Use snapshot read guard for Get operations

### CODE REVIEW

File: storage/src/tests/distributor/bucketdatabasetest.cpp [hunk 1]
```diff
@@ -170,8 +170,15 @@ BucketDatabaseTest::doFindParents(const std::vector<document::BucketId>& ids,
     }
 
     std::vector<BucketDatabase::Entry> entries;
+    // TODO remove in favor of only read guard once legacy DB usage has been ported over
     db().getParents(searchId, entries);
 
+    std::vector<BucketDatabase::Entry> checked_entries;
+    db().acquire_read_guard()->find_parents_and_self(searchId, checked_entries);
+    if(entries != checked_entries) {
```
Comment (geirst) [inline storage/src/tests/distributor/bucketdatabasetest.cpp]: Nitpick: Missing space after 'if'

**Review by geirst — APPROVED** [review]
👍

**Comment by geirst** [comment]
@toregge FYI

File: storage/src/tests/distributor/bucketdatabasetest.cpp (hunk 1 above)
Comment (vekterli) [reply storage/src/tests/distributor/bucketdatabasetest.cpp]: Fixed

**Pull request merged**